crates/pathfinder/src/state/sync/l2.rs

```diff
@@ -28,6 +30,65 @@ pub struct Timings {
     pub class_declaration: Duration,
 }
 
+/// A cache containing the last `N` downloaded class hashes, meant to minimize the
+/// re-downloading of class definitions in cases where a block might be re-processed
+/// e.g. after a reorg, or when a pending block becomes the new latest block.
+type ClassCache = LruCache<ClassHash, ()>;
+
+/// A cache containing the last `N` blocks in the chain. Used to determine reorg extents
+/// and ensure the integrity of new blocks.
+pub struct BlockChain {
+    head: BlockNumber,
+    tail: BlockNumber,
+
+    map: HashMap<BlockNumber, (BlockHash, StateCommitment)>,
+}
+
+impl BlockChain {
+    pub fn reset_to_genesis(&mut self) {
+        self.map.drain();
+        self.head = BlockNumber::default();
+        self.tail = BlockNumber::default();
+    }
+
+    pub fn with_capacity(
+        capacity: usize,
+        blocks: Vec<(BlockNumber, BlockHash, StateCommitment)>,
+    ) -> Self {
+        let skip = blocks.len().saturating_sub(capacity);
+        let blocks = &blocks[skip..];
+
+        let head = blocks.first().map(|b| b.0).unwrap_or_default();
+        let tail = blocks.last().map(|b| b.0).unwrap_or_default();
+
+        let mut map = HashMap::with_capacity(capacity);
+        map.extend(blocks.iter().cloned().map(|(a, b, c)| (a, (b, c))));
+
+        Self { head, tail, map }
+    }
+
+    pub fn block_hash<'a>(
+        &'a self,
+        block: &BlockNumber,
+    ) -> Option<&'a (BlockHash, StateCommitment)> {
+        self.map.get(block)
+    }
+
+    pub fn push(&mut self, number: BlockNumber, hash: BlockHash, commitment: StateCommitment) {
```

Comment:
Either I'm not understanding the code correctly or push contradicts with the ctor (`with_capacity`).

- if the order of blocks passed to `with_capacity` is ascending, then `tail >= head`:
```
H _ _ _ T
n increases --->
```
- but then this means that `head` and `tail` in this function should be swapped
```
// in normal case (number == T+1)

H _ _ _ T   // was
x H _ _ _ T // is

removal range = number..=T == {}
remove(H); H+=1;
T=number; insert(number, ... );

// for reorg (number <= T):

H _ _ _ T // was
H _ T x x // is

removal range = number..=T
T=number; insert(number, ... );
```

Reply:
Thanks - sometimes I really wonder if I can actually code 🤦

Reply:
Any other tests you'd like to add there?

Reply:
Nope, thanks :bow: